### DISCUSSION TITLE
Upstream_list not contain all dependencies

### DISCUSSION THREAD

**Discussion by YvYh** [apache/airflow#22949] [Q&A] [answered]
We have recently add a PythonOperator task into our Dag which depends on many other tasks and with tigger_rule equals to "all_success".

The Graph View and the Tree View show all dependencies correctly, but in its Task Instance Details it has only one task in "upstream_list". The consequnce is it will run immediatly after this task finish even though others dependencies are not yet finished. 

However in the same Dag, another BashOperator(which exists and works for a long time) alos with many upstreams doesn't have this problem, it can show all dependencies in "upstream_list" and run as designed.

Any idea about this point?

P.S. we are in version 1.10.15

**Comment by potiuk** [comment]
Impossible to say without seeing your DAG. What you describe does not remind any problems. 

BTW. Upgrade to Airflow 2. Airflow 1.10 is End of Life for almost 8 months and a) you will likely not get much support and help for it any more b) it will not receive even critical fixes so if there will be any problem found there. you will have to upgrade anyway. I recommend migration to 2.2.5.

**Comment by YvYh** [comment, marked as answer]
Well after checking, finally we found out that the problem is caused by the loop function in Dag script.

Actually we use a loop to create tasks as well as their relationships. As in each iteration it will create this PythonOperator and always with the **same id**, in graph or tree view it considers this task has serval dependences but when Dag runs it takes only the relation defined in the last iteration. That’s why we saw only one upstream in Task Instance.

The solution is to move the definition of this PythonOperator out of the loop but keep the definition of dependencies inside the loop. This well resolves the problem of running order, the final task won’t start until all dependences are finished.

**Reply by potiuk** [reply]
I see. Yep. That makes sense.
